mcs/tools/mono-api-diff/mono-api-diff.cs

```diff
@@ -53,35 +56,92 @@ static int Main (string [] args)
 				return 1;
 			}
 
-			XMLAssembly ms = CreateXMLAssembly (extra [0]);
-			XMLAssembly mono = CreateXMLAssembly (extra [1]);
-			XmlDocument doc = ms.CompareAndGetDocument (mono);
+			TextWriter outputStream = null;
+			try {
+				if (!string.IsNullOrEmpty (output))
+					outputStream = new StreamWriter (output);
 
-			StreamWriter outputStream = null;
-			if (!string.IsNullOrEmpty (output))
-				outputStream = new StreamWriter (output);
+				ApiDiff.Generate(extra[0], extra[1], outputStream ?? Console.Out);
+			} catch (Exception e) {
+				Console.WriteLine (e);
+				return 1;
+			} finally {
+				outputStream?.Dispose ();
+			}
+			return 0;
+		}
+	}
+#endif
+
+	public static class ApiDiff
+	{
+		public static void Generate (string firstInfo, string secondInfo, TextWriter outStream)
+		{
+			if (firstInfo == null)
+				throw new ArgumentNullException (nameof (firstInfo));
+			if (secondInfo == null)
+				throw new ArgumentNullException (nameof (secondInfo));
+
+			XMLAssembly ms = CreateXMLAssembly (firstInfo);
+			XMLAssembly mono = CreateXMLAssembly (secondInfo);
+
+			Generate (ms, mono, outStream);
+		}
 
-			using (XmlTextWriter writer = new XmlTextWriter (outputStream ?? Console.Out)) {
+		public static void Generate (Stream firstInfo, Stream secondInfo, TextWriter outStream)
+		{
+			if (firstInfo == null)
+				throw new ArgumentNullException (nameof (firstInfo));
+			if (secondInfo == null)
+				throw new ArgumentNullException (nameof (secondInfo));
+
+			XMLAssembly ms = CreateXMLAssembly (firstInfo);
+			XMLAssembly mono = CreateXMLAssembly (secondInfo);
+
+			Generate (ms, mono, outStream);
+		}
+
+		static void Generate (XMLAssembly first, XMLAssembly second, TextWriter outStream)
+		{
+			if (first == null)
+				throw new ArgumentNullException (nameof (first));
+			if (second == null)
+				throw new ArgumentNullException (nameof (second));
+			if (outStream == null)
+				throw new ArgumentNullException (nameof (outStream));
+
+			XmlDocument doc = first.CompareAndGetDocument (second);
+
+			using (XmlTextWriter writer = new XmlTextWriter (outStream)) {
 				writer.Formatting = Formatting.Indented;
 				doc.WriteTo (writer);
 			}
-
-			return 0;
 		}
 
 		static XMLAssembly CreateXMLAssembly (string file)
+		{
+			using (var stream = File.OpenRead(file)) {
+				return CreateXMLAssembly (stream);
+			}
+		}
+
+		static XMLAssembly CreateXMLAssembly (Stream stream)
 		{
 			XmlDocument doc = new XmlDocument ();
-			doc.Load (File.OpenRead (file));
+			doc.Load (stream);
 
 			XmlNode node = doc.SelectSingleNode ("/assemblies/assembly");
 			XMLAssembly result = new XMLAssembly ();
+#if !EXCLUDE_DRIVER
 			try {
+#endif
 				result.LoadData (node);
+#if !EXCLUDE_DRIVER
 			} catch (Exception e) {
-				Console.Error.WriteLine ("Error loading {0}: {1}\n{2}", file, e.Message, e);
+				Console.Error.WriteLine ("Error loading {0}: {1}\n{2}", (stream as FileStream)?.Name ?? "<stream>", e.Message, e);
```

Comment:
What do you think about abstracting the output handling into a class that either writes to`Console.Error`/`Console.Out` (for .exe) or a StringBuilder (for .dll)?

That way you can get ahold of the log messages even in a library build and we don't need to litter the code with `EXCLUDE_DRIVER` ifdefs.